### PR TITLE
Testing section

### PR DESCRIPTION
**Pull request by katrinsharp**
Adding testing page to documentation explaining unit testing in cross project as well as how to configure integration testing in cross project. Let me know what you think. /cc @ochrons

### CODE REVIEW

File: .htaccess [hunk 1]
```diff
@@ -12,6 +12,7 @@ redirect 301 /doc/sbt/run.html {{ BASE_PATH }}/doc/project/building.html
 redirect 301 /doc/sbt/depending.html {{ BASE_PATH }}/doc/project/dependencies.html
 redirect 301 /doc/sbt/js-envs.html {{ BASE_PATH }}/doc/project/js-environments.html
 redirect 301 /doc/sbt/cross-building.html {{ BASE_PATH }}/doc/project/cross-build.html
+redirect 301 /doc/sbt/testing.html {{ BASE_PATH }}/doc/project/testing.html
```
Comment (gzm0) [inline .htaccess]: I don't think this is needed. IIUC, these are just so that links to the old site still work (a comment would have been useful indeed).

**Comment by katrinsharp** [comment]
@gzm0 done - link to test section removed + comment about the file added

**Comment by gzm0** [comment]
> @gzm0 done - link to test section removed + comment about the file added

Now I hope what I said is actually the case :P

**Comment by katrinsharp** [comment]
@gzm0 I hope that someone will be able to approve/disapprove that, if not, then you're right  anyways :)

**Comment by sjrd** [comment]
> IIUC, these are just so that links to the old site still work (a comment would have been useful indeed).

Yes, that's correct. We also add redirects when moving pages around, of course. So the comment, as written, is wrong.

But there should not be a new entry in redirects for a new page like this.

File: .htaccess [hunk 2]
```diff
@@ -1,5 +1,12 @@
 ---
 ---
+
+{% comment %}
+Those are links to the old site so it will still work. 
+Change it ONLY if you deal with old site!
+{% endcomment %}
+
+
```
Comment (sjrd) [inline .htaccess]: The appropriate comment would be simply:

> Redirects for pages that were moved around.

**Comment by sjrd** [comment]
I'll review the actual content later. I'm on a clock, right now.

File: .htaccess [hunk 3]
```diff
@@ -1,5 +1,12 @@
 ---
 ---
+
+{% comment %}
+Update this file if you have an entry that you are moving around.
+Don't update this file if you are adding a new entry.
+{% endcomment %}
+
+
```
Comment (sjrd) [inline .htaccess]: Double blank line. (should be only 1)

File: doc/project/testing.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
+---
+layout: doc
+title: Testing
+---
+
+In this section we will discuss how to test Scala.js code managed as [Full cross project](./cross-build.html). All names for example subproject name such as `fooJVM` are taken from [cross compile example](https://github.com/scala-js/scalajs-cross-compile-example).
+
+
```
Comment (sjrd) [inline doc/project/testing.md]: Double blank line.

File: doc/project/testing.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+     |   +- src/main/scala
+     |   +- src/test/scala
+     +- js
+     |   +- src/main/scala
+     |   +- src/test/scala
+     +- shared
+         +- src/main/scala
+         +- src/test/scala
+
+Similarly to `{shared|js|jvm}/src/main/scala` folders that contain your code, there can be also `{shared|js|jvm}/src/test/scala` folders that would contain your tests. Tests that verify correctness of shared code should go to `shared` test folder, while tests that check Scala JVM code or Scala.js code only should go to `jvm` or `js` test folders respectively. When you call `sbt> test` two iterations of tests will run:
```
Comment (sjrd) [inline doc/project/testing.md]: Try to cut long lines in the .md file so that they do not exceed 120 characters.

> When you call `sbt> test` two iterations of tests will run:

This is only true if there is a root project which `aggregate`s the JVM and JS parts. This should be explicit.

File: doc/project/testing.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
++ Iteration 1: all tests residing in `shared` and `jvm` will be executed testing Scala JVM code.
++ Iteration 2: all tests residing in `shared` and `js` will be executed testing Scala.js code.
```
Comment (sjrd) [inline doc/project/testing.md]: We use `*` instead of `+` for bulleted lists.

File: doc/project/testing.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+If you are interested in running only one of those iterations you could call `sbt> fooJVM/test` or `sbt> fooJS/test` accordingly.
```
Comment (sjrd) [inline doc/project/testing.md]: I would *first* present the logic to test separately. Then just remind that with an aggregate project, calling `test` will just run `fooJVM/test` and `fooJS/test`.

File: doc/project/testing.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
++ Iteration 1: all tests residing in `shared` and `jvm` will be executed testing Scala JVM code.
++ Iteration 2: all tests residing in `shared` and `js` will be executed testing Scala.js code.
+
+If you are interested in running only one of those iterations you could call `sbt> fooJVM/test` or `sbt> fooJS/test` accordingly. 
+
+## Integration testing
+
+Configuring regular non-Scala.js sbt project to have `it:test` is a well described task, see [documentation](http://www.scala-sbt.org/0.12.4/docs/Detailed-Topics/Testing.html#integration-tests) for more details. In `CrossProject` it won't work out-of-the-box because by default Scala JVM and Scala.js `it` configurations won't be aware of `shared/src/it/scala` folder and additionally Scala.js `it` configuration by default does not contain `ScalaJSClassLoader` definition required for running Scala.js tests. To fix that you need to add the following settings to crossProject definition along with `IntegrationTest` settings:
```
Comment (sjrd) [inline doc/project/testing.md]: * Configuring <ins>a</ins> regular non-Scala.js
* Use the current URL to the documentation: http://www.scala-sbt.org/0.13/docs/Testing.html#Integration+Tests
* In <ins>a</ins> `crossProject`.
* The second sentence is very long. Try to insert a period somewhere.
* settings to <ins>the</ins> `crossProject`

File: doc/project/testing.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+{% highlight scala %}
+lazy val cross = crossProject.in(file(".")).
+  ...
+  // adding it configuration
+  .configs( IntegrationTest )
```
Comment (sjrd) [inline doc/project/testing.md]: The coding style used in our documentation (and sbt's) is to put the `.` at the end of the previous line. Also no spaces inside parentheses.

File: doc/project/testing.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+If you are interested in running only one of those iterations you could call `sbt> fooJVM/test` or `sbt> fooJS/test` accordingly. 
+
+## Integration testing
+
+Configuring regular non-Scala.js sbt project to have `it:test` is a well described task, see [documentation](http://www.scala-sbt.org/0.12.4/docs/Detailed-Topics/Testing.html#integration-tests) for more details. In `CrossProject` it won't work out-of-the-box because by default Scala JVM and Scala.js `it` configurations won't be aware of `shared/src/it/scala` folder and additionally Scala.js `it` configuration by default does not contain `ScalaJSClassLoader` definition required for running Scala.js tests. To fix that you need to add the following settings to crossProject definition along with `IntegrationTest` settings:
+
+{% highlight scala %}
+lazy val cross = crossProject.in(file(".")).
+  ...
+  // adding it configuration
```
Comment (sjrd) [inline doc/project/testing.md]: adding <ins>the</ins> `it` configuration

Also add backticks around "it" here and also below. Being such a common word in English, it makes sentences hard to parse without the backticks, because instinctively you think about the pronoun "it" rather than the configuration `it`.

File: doc/project/testing.md [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  .configs( IntegrationTest )
+  // adding it tasks
+  .settings( Defaults.itSettings : _*)
```
Comment (sjrd) [inline doc/project/testing.md]: No space after `(` nor after `:`

File: doc/project/testing.md [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    CrossType.Full.sharedSrcDir(baseDirectory.value, "it").toSeq)
+  // adding ScalaJSClassLoader to js configuration
+  .jsSettings(inConfig(IntegrationTest)(ScalaJSPluginInternal.scalaJSTestSettings):_*) 
+{% endhighlight %}
+
+Now you can put tests in `{shared|jvm|js}/src/it/scala` and they will run when you call `sbt> it:test`. Similarly to `test`, there is also two iterations and they can be executed separately by explicitely stating jvm or js project name, for example `sbt> fooJS/it:test`.
+
+## Testing frameworks
+
+List of test frameworks compatible with Scala.js can be found [here](../../libraries/testing.html).
```
Comment (sjrd) [inline doc/project/testing.md]: <ins>A</ins> list of test<ins>ing</ins> frameworks

File: doc/project/testing.md [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+Note: Don't forget to mark a test framework SBT dependency as `test, it` if you have both unit and integration tests.
```
Comment (sjrd) [inline doc/project/testing.md]: The sbt documentation suggests there must not be a space in `test,it`

**Comment by sjrd** [comment]
That's all.

**Comment by katrinsharp** [comment]
@sjrd I believe I addressed all your comments in latest push. Thanks

**Comment by sjrd** [comment]
OK could you squash all the commits into 1, please?

**Comment by katrinsharp** [comment]
Squashed to 1 commit per @sjrd request.

**Comment by sjrd** [comment]
LGTM
